Allow deploy Admin VMs and VRs in disabled zones/pods/clusters

GabrielBrascher opened the pull request:
## Description
<!--- Describe your changes in detail -->
This proposal allows Virtual Routers to be deployed on disabled zones/pods/clusters. It also gives the opportunity to deploy Admin VMs to be deployed on disabled resources.

The default configuration is set to **false** for deploying Admin VMs on disabled resources and deploying system VMs on disabled resources. Both, `allow.router.on.disabled.resources` and `allow.admin.vm.on.disabled.resources`, are configurable and do not require management server restart when updated.

The main goal of this implementation is to give flexibility and total control for system admins to decide either if they want or not to deploy System and/or Admin VMs in disabled resources. This implementation was encouraged after an issue related with a disabled Pod; it was not possible to reset networks in the Zone due to the fact that the VRs running on the disabled Pods would fail to be re-deployed.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

andrijapanicsb commented:
@GabrielBrascher can we please rename the setting from "allow.admin.vm.disabled.resources" to "allow.admin.vm.ON.disabled.resources" (word "on" added) - so they make more sense for new users? (same for system.vm). thx

GabrielBrascher commented:
@andrijapanicsb sure, I will add "on" at both parameters, thanks!

rhtyd commented on engine/components-api/src/main/java/com/cloud/deploy/DeploymentPlanningManager.java:
```diff
@@ -24,6 +26,12 @@
+    static final ConfigKey<Boolean> allowSytemVmInDisabledResource = new ConfigKey<Boolean>("Advanced", Boolean.class, "allow.system.vm.disabled.resource", "true",
+            "Allow deploying System VMs in disabled Clusters, Pods, and Zones (does not require restart)", true);
+
+    static final ConfigKey<Boolean> allowAdminVmInDisabledResource = new ConfigKey<Boolean>("Advanced", Boolean.class, "allow.admin.vm.disabled.resources", "false",
```
@GabrielBrascher Can both be made false to ensure the current behaviour? (we don't want to surprise users and admins after upgrade)

GabrielBrascher replied:
@rhtyd I see your point. I understand that we need to avoid such surprises when upgrading. Do you know a way to update it without surprising users? I ask this because, from my perspective, such parameter should be set to true as default for system VMs.

Just to provide some context on why this PR has been created. These changes are tackling an issue regarding cases as the following:

1. each pod requires a VR for a network 'A';
2. network 'A' has been restarted and all VRs of that network have been re-deployed;
3. VRs fail to be deployed on disabled Pods (that still running production VMs);
4. network restarting fails for disabled Pods and cannot update network configurations that we had expected to be updated on the VR.
5. production VMs running on disabled Pods have no longer the VR services.

The idea is that we should, by default, keep deploying such VMs to provide all the services also for legacy environments.

I see a disabled resource as a legacy environment where one shall not deploy new production VMs; however, such an environment still need to provide the basic services for VMs that have not been destroyed/migrated. Thus, system VMs (CCPVM, SSVM, VRs) are still necessary to be deployed/re-deployed to keep serving the VMs that have not yet been migrated from the disabled resource.

Said that, if you still think that this should be set to false by default, I can add info at error messages that point to this parameter when system VMs are not deployed on disable resources. At least making the user/admin aware of this option.

GabrielBrascher replied:
@rhtyd I've added a commit addressing your comment, thx!

rhtyd commented:
@GabrielBrascher can you check and fix build failure, see Travis

GabrielBrascher commented:
@rhtyd Fixed it; a line trailing on the test class that raised the check style issue. Thanks!

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✔debian. JID-307

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-403)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 32983 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3600-t403-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_affinity_groups_projects.py
Intermittent failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermittent failure detected: /marvin/tests/smoke/test_projects.py
Smoke tests completed. 75 look OK, 2 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_DeployVmAntiAffinityGroup_in_project | `Error` | 1.32 | test_affinity_groups_projects.py
test_09_project_suspend | `Error` | 2.43 | test_projects.py
test_10_project_activation | `Error` | 2.31 | test_projects.py

GabrielBrascher commented:
Hi @andrijapanicsb and @rhtyd,

Parameters were updated with the _ON_: "allow.system.vm.on.disabled.resource" and "allow.admin.vm.on.disabled.resources".

Both parameters are set as false by default. Thus, this commit does not change CloudStack behavior and should not impact on automated tests neither surprise users and admins after upgrade.

Thanks for the reviews!

PaulAngus commented:
If I understand correctly, your use case only applies when you've disabled all compute resources in a pod (either independently or as one) in a basic network. 
A VR can be deployed in any pod to facilitate a network in advanced networks.
So this is a bit of a corner case.

Therefore setting the default to false (carry on as before) would seem the most sensible solution to me also.

GabrielBrascher commented:
That is the idea @PaulAngus.

Quite a corner case that this global settings will help with; thus, leaving it disabled (false) by default. However, this commit turns out to be very handy in specific situations such as disabled resources (luster, pod or zone)that will be deprecated but still need System VMs.

andrijapanicsb commented on server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java:
```diff
@@ -548,6 +547,78 @@ public DeployDestination planDeployment(VirtualMachineProfile vmProfile, Deploym
+    }
+
+    /**
+     * Returns the value of the ConfigKey 'allow.system.vm.disabled.resource'. This method allows mocking and testing with the respective ConfigKey parameter.
```
"Returns the value of the ConfigKey 'allow.system.vm.ON.disabled.resource" - since name was changed.

andrijapanicsb commented on server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java:
```diff
@@ -548,6 +547,78 @@ public DeployDestination planDeployment(VirtualMachineProfile vmProfile, Deploym
+    }
+
+    /**
+     * Returns the value of the ConfigKey 'allow.admin.vm.disabled.resources'. This method allows mocking and testing with the respective ConfigKey parameter.
```
"Returns the value of the ConfigKey 'allow.admin.vm.ON.disabled.resources'" since the name was changed

GabrielBrascher replied:
@andrijapanicsb updated both Javadoc blocks :+1:

andrijapanicsb commented:
@blueorangutan package

blueorangutan commented:
@andrijapanicsb a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✔debian. JID-429

andrijapanicsb commented:
@blueorangutan test

blueorangutan commented:
@andrijapanicsb a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-582)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 34071 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3600-t582-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_affinity_groups_projects.py
Intermittent failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermittent failure detected: /marvin/tests/smoke/test_projects.py
Smoke tests completed. 75 look OK, 2 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_DeployVmAntiAffinityGroup_in_project | `Error` | 1.22 | test_affinity_groups_projects.py
test_09_project_suspend | `Error` | 2.29 | test_projects.py
test_10_project_activation | `Error` | 2.22 | test_projects.py

andrijapanicsb commented:
@blueorangutan test

blueorangutan commented:
@andrijapanicsb a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

DaanHoogland commented on engine/schema/src/main/java/com/cloud/host/dao/HostDaoImpl.java:
```diff
@@ -431,13 +431,24 @@ public long countBy(long clusterId, ResourceState... states) {
+        SearchCriteria<HostVO> sc = scHostsFromZoneUpRouting(id);
+        sc.setParameters("resourceState", ResourceState.Disabled);
+        return listBy(sc);
+    }
```
It makes sense to unify this to listByDataCentyerIdAndState() internally, no?

DaanHoogland replied:
ok, not what i meant but looks good as well.  ( i said internally ;)

GabrielBrascher commented:
@DaanHoogland I just added method listByDataCenterIdAndState(long, ResourceState) to  HostDao. Can you please take a look? Thanks!

blueorangutan commented:
<b>Trillian test result (tid-679)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 39825 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3600-t679-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_affinity_groups_projects.py
Intermittent failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermittent failure detected: /marvin/tests/smoke/test_projects.py
Smoke tests completed. 74 look OK, 3 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_DeployVmAntiAffinityGroup_in_project | `Error` | 1.26 | test_affinity_groups_projects.py
test_02_vpc_privategw_static_routes | `Failure` | 259.26 | test_privategw_acl.py
test_03_vpc_privategw_restart_vpc_cleanup | `Failure` | 248.39 | test_privategw_acl.py
test_09_project_suspend | `Error` | 2.30 | test_projects.py
test_10_project_activation | `Error` | 2.22 | test_projects.py

rhtyd commented:
@GabrielBrascher Travis was failing all jobs, rekicked by closing/opening the PR. Can you also check if there are any unit test or build failures?

GabrielBrascher commented:
@rhtyd I re-kicked with a force push. All tests are passing; however, there was a trailing space left that was failing on checkstyle. I've just fixed it :-)

GabrielBrascher commented:
@rhtyd @DaanHoogland @andrijapanicsb It looks that Travis CI build has some env issues.

```
WARNING: Requirement 'tools/marvin/dist/Marvin-*.tar.gz' looks like a filename, but the file does not exist
Processing ./tools/marvin/dist/Marvin-*.tar.gz
ERROR: Could not install packages due to an EnvironmentError: [Errno 2] No such file or directory: '/home/travis/build/apache/cloudstack/tools/marvin/dist/Marvin-*.tar.gz'

WARNING: You are using pip version 19.1.1, however version 20.0.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
The command "./tools/travis/install.sh" failed and exited with 1 during .
```

DaanHoogland commented:
@andrijapanicsb are all your objections met?

andrijapanicsb commented:
I believe so, @DaanHoogland

DaanHoogland commented:
than, can you approve please @andrijapanicsb

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✔debian. JID-721

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✔debian. JID-766

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-912)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 31601 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3600-t912-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_affinity_groups_projects.py
Intermittent failure detected: /marvin/tests/smoke/test_projects.py
Smoke tests completed. 77 look OK, 2 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_DeployVmAntiAffinityGroup_in_project | `Error` | 1.48 | test_affinity_groups_projects.py
test_09_project_suspend | `Error` | 2.43 | test_projects.py
test_10_project_activation | `Error` | 2.32 | test_projects.py

andrijapanicsb commented:
@blueorangutan test

blueorangutan commented:
@andrijapanicsb a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

andrijapanicsb commented:
@GabrielBrascher I gave this a brief test, and it seems it's not working for system VMs (SSVM/CPVM) or I'm doing something wrong:
- have both SSVM/CPVM running 
- disabled the only host in the zone/pod
- destroyed the SSVM, it get's recreated as expected
- set "allow.system.vm.on.disabled.resource" to "true"
- destroy the SSVM again - it doesn't get started (not even after mgmt server restart, though the setting says it's dynamic=1)

notice the missing "s" at the end of the "allow.system.vm.on.disabled.resource" vs " allow.admin.vm.on.disabled.resource**s**" 

Am I doing something wrong?

andrijapanicsb commented:
It does work fine for already existing VR - it can be started when allow.system.vm.on.disabled.resource=true

Also ok for user VMs (started by admin).

@GabrielBrascher I disabled only the host (not zone/pod/cluster) - is this by design/OK - and should you just update the description of the global settings to also say "allow .. on disabled hosts" ?

GabrielBrascher commented:
I will re-test it today and check the code and add the missing "s" as well. Thanks for the heads up @andrijapanicsb.

andrijapanicsb commented:
thx @GabrielBrascher.

It seems that cloud admin can only start his own userVM on disabled resources - i.e. Admin can't start another user's userVM. But admin can start another user's VR.
Is this by design?

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✔debian. JID-810

GabrielBrascher commented:
@andrijapanicsb thanks for raising those points. Sorry for the delay, it has been 4 months since I opened this PR and I needed to take a step back and look again to the use case, proposal, and implementation.

> It seems that cloud admin can only start his own user VM on disabled resources - i.e. Admin can't start another user's user VM. But admin can start another user's VR.
> Is this by design?

You are right, the goal of this implementation is that admins can start their own VMs and System VMs; however, user VMs should not be deployed on disabled resources. It tackles a very specific case that we've seen in deprecated clusters/pods.

>  I gave this a brief test, and it seems it's not working for system VMs (SSVM/CPVM)

That is indeed related to the fact that there is only one host and the host is disabled. I did some tests and it works if disabling the cluster/pod/zone.

> I disabled only the host (not zone/pod/cluster) - is this by design/OK - and should you just update the description of the global settings to also say "allow .. on disabled hosts" 

The initial idea was indeed to allow hosts as well; however, there were some issues that would put the implementation on another level of complexity. It would also be a very specific corner case where all hosts are disabled, therefore it would make sense to disable the cluster/pod/zone instead of all hosts. That is why I described PR with disabled _zones/pods/clusters_.

Additionally, from the implementation view: the code will mark storage pools as not suitable in case of local storage and disabled hosts. If there is no linkage of any suitable host (all hosts disabled) to any of the pools in the cluster, then the workflow considers the storage pool as "disabled" (not only hosts). Therefore, due to high complexity on solution for a very specific case, I decided to keep the implementation on zone/pod/cluster scope.

> notice the missing "s" at the end of the "allow.system.vm.on.disabled.resource" vs " allow.admin.vm.on.disabled.resources"

Going to fix that and also enhance documentation to let it clear that it works for disabled zone/pod/cluster :)

andrijapanicsb commented:
thx @GabrielBrascher - will retest once I'm available - I'm OK with your other comments/answers - thx!

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✔debian. JID-910

andrijapanicsb commented:
@blueorangutan test

blueorangutan commented:
@andrijapanicsb a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

andrijapanicsb commented:
@GabrielBrascher I gave this another test, where I had 2 clusters of 2 KVM hosts.
- admin VMs - all fine
- VRs - all fine
- CPVM/SSVM not fine as below, similar/multiple issues - the behaviour is identical as before this PR

### test disabled clusters only
- settings "allow.system.vm.on.disabled.resources" by default is set to false (please update the description of PR - it is NOT set to "true" for systemVMs)
- Disabled both clusters (not zone/pod/host)
- stop or destroy SSVM - and it still gets started/new one created and started even though both clusters are disabled.

### test disabled pod
- same setup as when disabling the cluster, just now we are disabling the POD only - identical behaviour for SSVM/CPVM

### test disable zone (no changes to behaviour from before)
- same setup as when disabling the clusters, but now we are disabling the zone only
- a stopped SSVM will NOT be started autoamtically, but can be started manually (this si the same as before, I've briefly tested/confirmed)
- a destroyed SSVM will NOT be recreated - so this is fine (as before)

### test disable zone, but allow.system.vm.on.disabled.resources=TRUE
- disable the zone
- identical behaviour as when allow.system.vm.on.disabled.resources=FALSE - i.e. as before this PR

So this is not working for SSVM/CPVM definitively - if you don't want to extend the scope to cover CPVM/SSVM properly (which I'm fine with actually...) then perhaps confirm this is in line with your code (what I tested here) and rename just the setting to "...vr...." instead of "..system.vm....." - as it only affects VR per my testing.

One thing to keep in mind, if only one cluster is disabled, the VM can be always started in another cluster irrelevant of these 2 settings you introduced here (funny enough - the volume is **silently migrated** to the primary storage pool of that other cluster and VM is started now in the other cluster which is still enabled) - so this is not a valid test - you need to disable all resources (i.e. all clusters, like I did in order to see if it works as expected)

Thoughts?

blueorangutan commented:
<b>Trillian test result (tid-1063)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 27893 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3600-t1063-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_affinity_groups_projects.py
Intermittent failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermittent failure detected: /marvin/tests/smoke/test_projects.py
Smoke tests completed. 77 look OK, 3 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_DeployVmAntiAffinityGroup_in_project | `Error` | 1.18 | test_affinity_groups_projects.py
test_02_vpc_privategw_static_routes | `Failure` | 177.91 | test_privategw_acl.py
test_03_vpc_privategw_restart_vpc_cleanup | `Failure` | 181.90 | test_privategw_acl.py
test_04_rvpc_privategw_static_routes | `Failure` | 242.48 | test_privategw_acl.py
test_09_project_suspend | `Error` | 2.26 | test_projects.py
test_10_project_activation | `Error` | 2.22 | test_projects.py

GabrielBrascher commented:
@andrijapanicsb once again, thank you for all the feedback.

Just to summarize:

(i) the good news first :slightly_smiling_face:: we both had success starting VR and Admin VMs on disabled Zone/Pod/Cluster.
(ii) However, CPVM and SSVM fail to start when the Zone is disabled. In any other situation, they work fine.

I understood what is causing the issue that you reported. SSVMs and CPVMs are started automatically by ACS when the zone is enabled, differently from VRs that can be deployed via another way (e..g deploying a VM on the given network).

I will take a quick look searching for answers. If I fail to solve this today, tomorrow I will change the scope of this PR to VRs and Admin VMs.

andrijapanicsb commented:
Sure, need to support the next VP ;)

GabrielBrascher commented:
I had to update it to the current master due to some failures. Rebased and pushed again, that is why the ugly `force-pushed` : (

GabrielBrascher commented:
Last commit "Documentation updated to allowing deployment of VRs on disabled zones." updated the code to fit with the SSVM/CPVM behavior on disabled zones.

andrijapanicsb commented:
@GabrielBrascher sorry, is this ready now for retesting i.e.- anything else to put in? thx

andrijapanicsb commented:
@blueorangutan package

blueorangutan commented:
@andrijapanicsb a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

GabrielBrascher commented:
@andrijapanicsb this one is ready now. Thanks!

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✔debian. JID-953

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-1134)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 28743 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3600-t1134-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_affinity_groups_projects.py
Intermittent failure detected: /marvin/tests/smoke/test_projects.py
Intermittent failure detected: /marvin/tests/smoke/test_router_dhcphosts.py
Smoke tests completed. 78 look OK, 2 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_DeployVmAntiAffinityGroup_in_project | `Error` | 1.20 | test_affinity_groups_projects.py
test_09_project_suspend | `Error` | 2.24 | test_projects.py
test_10_project_activation | `Error` | 2.20 | test_projects.py

DaanHoogland commented:
lots of errors in travis and some in smoke tests. Maybe you can rebase your code @GabrielBrascher ?

GabrielBrascher commented:
@DaanHoogland I am going to check those test and rebase as well.

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✔debian. JID-975

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-1157)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 27660 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3600-t1157-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_affinity_groups_projects.py
Intermittent failure detected: /marvin/tests/smoke/test_projects.py
Smoke tests completed. 78 look OK, 2 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_DeployVmAntiAffinityGroup_in_project | `Error` | 1.21 | test_affinity_groups_projects.py
test_09_project_suspend | `Error` | 2.27 | test_projects.py
test_10_project_activation | `Error` | 2.21 | test_projects.py

GabrielBrascher commented:
Still got the erros after rebasing to master. I am going to take another look at it.

DaanHoogland commented:
ping @GabrielBrascher will you make it to get this done by friday?

GabrielBrascher commented:
@DaanHoogland the plan is to have it solved in time, I will let you know if not possible to have it figured out before Friday.

andrijapanicsb commented:
ping @GabrielBrascher freeze incoming in 13-14h :) any update? And there is a conflict.

andrijapanicsb commented:
Last ping @GabrielBrascher  :) - any change to resolve this?
Otherwise I shall move it to 4.15?

andrijapanicsb commented:
Moving to 4.15

GabrielBrascher commented:
Sorry for the radio silence, I was sick last Friday. Better now at least ;-)

GabrielBrascher commented:
@andrijapanicsb @DaanHoogland I just rebased and fixed all conflicts, branch good for master/4.15 now.

Note: I have executed a batch of manual tests on a 4.15-SNAPSHOT Zone. All looking good!

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos7 ✖debian. JID-1435

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos7 ✔debian. JID-1437

GabrielBrascher commented:
@andrijapanicsb @rhtyd @DaanHoogland is this one ready for merge?
Let me know if there is anything else that I should address on this PR :-)

DaanHoogland commented:
two oks so it should be , but it has been laying in the cellar to mature for quite a while, so let me check how stale it is:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos7 ✔centos8 ✔debian. JID-2063

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

DaanHoogland commented:
@GabrielBrascher it keeps failing as our lab is to busy. did you run integration tests on this? (cc @PaulAngus )

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos7 ✔centos8 ✔debian. JID-2088

GabrielBrascher commented:
@blueorangutan package

GabrielBrascher commented:
Fixed conflict with Master.

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos7 ✔centos8 ✔debian. JID-2116

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-2911)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 39748 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3600-t2911-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_affinity_groups_projects.py
Intermittent failure detected: /marvin/tests/smoke/test_diagnostics.py
Intermittent failure detected: /marvin/tests/smoke/test_kubernetes_clusters.py
Intermittent failure detected: /marvin/tests/smoke/test_projects.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermittent failure detected: /marvin/tests/smoke/test_hostha_kvm.py
Smoke tests completed. 81 look OK, 4 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_DeployVmAntiAffinityGroup_in_project | `Error` | 1.35 | test_affinity_groups_projects.py
test_09_project_suspend | `Error` | 2.49 | test_projects.py
test_10_project_activation | `Error` | 2.36 | test_projects.py
test_03_create_redundant_VPC_1tier_2VMs_2IPs_2PF_ACL_reboot_routers | `Failure` | 300.55 | test_vpc_redundant.py
test_hostha_kvm_host_fencing | `Error` | 175.95 | test_hostha_kvm.py

DaanHoogland commented:
not sure what still needs to be done here... @GabrielBrascher @andrijapanicsb @rhtyd ?

DaanHoogland commented:
rebuilding on latest changes
@blueorangutan package

blueorangutan commented:
@DaanHoogland a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos7 ✖centos8 ✖debian. JID-2216

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos7 ✔centos8 ✔debian. JID-2234

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-3031)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 33980 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3600-t3031-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_affinity_groups_projects.py
Intermittent failure detected: /marvin/tests/smoke/test_kubernetes_clusters.py
Intermittent failure detected: /marvin/tests/smoke/test_projects.py
Smoke tests completed. 82 look OK, 3 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_DeployVmAntiAffinityGroup_in_project | `Error` | 1.36 | test_affinity_groups_projects.py
test_03_deploy_and_upgrade_kubernetes_cluster | `Failure` | 248.35 | test_kubernetes_clusters.py
test_09_project_suspend | `Error` | 2.51 | test_projects.py
test_10_project_activation | `Error` | 2.37 | test_projects.py

DaanHoogland commented:
@GabrielBrascher the test_project errors are consistent, can you have a look?

GabrielBrascher commented:
I am going to take a look at it today. Thanks for the ping @DaanHoogland!

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos7 ✖centos8 ✔debian. JID-2289

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-3094)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 35321 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3600-t3094-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_affinity_groups_projects.py
Intermittent failure detected: /marvin/tests/smoke/test_projects.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Smoke tests completed. 84 look OK, 2 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_DeployVmAntiAffinityGroup_in_project | `Error` | 1.37 | test_affinity_groups_projects.py
test_09_project_suspend | `Error` | 2.48 | test_projects.py
test_10_project_activation | `Error` | 2.33 | test_projects.py

GabrielBrascher commented:
It still the same error on blueoragutan tests even after merge/rebase with latest master. Let me check this again.

DaanHoogland commented:
ping @GabrielBrascher will you have time to address this (be)for(e) 4.15?

GabrielBrascher commented on server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java:
```diff
@@ -574,6 +580,80 @@ private boolean getDisplayStorageFromVmProfile(VirtualMachineProfile vmProfile)
+
+            VMInstanceVO vm = _vmInstanceDao.findById(vmProfile.getId());
+            AccountVO owner = accountDao.findById(vm.getAccountId());
+            if (owner.getRoleId() == ADMIN_ACCOUNT_ROLE_ID && isAdminVmDeployableInDisabledResources()) {
```
@DaanHoogland the problem raised by the tests is due to the fact that the role ID of a project is null; thus, these tests are failing.

Tests that failed:
/marvin/tests/smoke/test_affinity_groups_projects.py
/marvin/tests/smoke/test_kubernetes_clusters.py
/marvin/tests/smoke/test_projects.py
I am going to take a look at a fix for it, not sure if this will be on 4.15 then. Is there an ETA for 4.15?

DaanHoogland replied:
:) last month, @GabrielBrascher

GabrielBrascher commented on server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java:
```diff
@@ -574,6 +580,86 @@ private boolean getDisplayStorageFromVmProfile(VirtualMachineProfile vmProfile)
+            AccountVO owner = accountDao.findById(vm.getAccountId());
+            boolean isOwnerRoleIdAdmin = false;
+
+            if (owner != null && owner.getRoleId() != null && owner.getRoleId() == ADMIN_ACCOUNT_ROLE_ID) {
```
@DaanHoogland I fixed the issue raised on tests regarding VMs deployed on a project.

Manual tests are looking good now; thanks :-)

DaanHoogland commented:
@blueorangutan package

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos7 ✔centos8 ✔debian. JID-2367

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-3155)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 32507 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3600-t3155-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_privategw_acl.py
Smoke tests completed. 85 look OK, 1 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_04_rvpc_privategw_static_routes | `Failure` | 914.38 | test_privategw_acl.py

GabrielBrascher commented:
@blueorangutan test

DaanHoogland commented:
> @blueorangutan test

pretty sure this is not needed or possible, @GabrielBrascher

blueorangutan commented:
@DaanHoogland unsupported parameters provided. Supported mgmt server os are: `centos6, centos7, centos8, ubuntu`. Supported hypervisors are: `kvm-centos6, kvm-centos7, kvm-centos8, kvm-ubuntu, xenserver-71, xenserver-65sp1, vmware-67u3, vmware-65u2, vmware-60u2, vmware-55u3, xcpng76, xcpng80, xcpng81, xenserver-74, xcpng74`

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_multiplication_x: centos7 :heavy_multiplication_x: centos8 :heavy_multiplication_x: debian. SL-JID 382

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_multiplication_x: centos7 :heavy_multiplication_x: centos8 :heavy_multiplication_x: debian. SL-JID 398

blueorangutan commented:
Packaging result: :heavy_multiplication_x: centos7 :heavy_multiplication_x: centos8 :heavy_multiplication_x: debian. SL-JID 404

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: centos7 :heavy_check_mark: centos8 :heavy_check_mark: debian. SL-JID 412

shwstppr commented:
@blueorangutan test

blueorangutan commented:
@shwstppr a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-442)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 32310 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3600-t442-kvm-centos7.zip
Smoke tests completed. 88 look OK, 0 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

GabrielBrascher commented:
Thanks, @shwstppr! Tests are looking good :+1:

GabrielBrascher commented:
Again annoying with some _pinging for review_: @DaanHoogland @andrijapanicsb @PaulAngus @svenvogel @RodrigoDLopez @GutoVeronezi @weizhouapache @ravening @rhtyd & others

This one completed 1 year in April :cake: :birthday: :balloon: :laughing:. I appreciate any feedback.
Thanks in advance :-)

PaulAngus commented:
Have specific Marvin tests been added for this?  Otherwise, a BlueOrangutan run won't determine if the feature actually works as required i.e. manual tests need to be recorded.

andrijapanicsb commented:
Very valid point by Paul @GabrielBrascher  (have smoke tests been added for this?)
I will try to test this tomorrow manually (building an anv now)

GabrielBrascher commented:
@PaulAngus you are right. There has not been added smoke test.

@andrijapanicsb I really appreciate any tests.

DaanHoogland commented:
Do I need to re-review @GabrielBrascher ?

slavkap commented:
@GabrielBrascher, I did manual tests

**disabled Zone:**
deploy of a VM - works
recreate a VR - works
recreate of CPVM/SSVM - doesn't

**disabled Pod:**
deploy of a VM - works
recreate a VR - works
recreate of CPVM/SSVM - works

**disabled Cluster:**
deploy of a VM - works
recreate a VR - works
recreate of CPVM/SSVM - works

But I'm not sure how to test with the below set up to deploy a new VR without deploying a guest VM?

allow.admin.vm.on.disabled.resources = false
allow.router.on.disabled.resources = true

andrijapanicsb commented:
@slavkap you can restart the network (or VPC if it's a VPC) with "clean up" - it will first try to create a new VR, later destroy the old one - see if the new new VR ge's created immediately.

Btw, since you tested this manually, and since I did not have time today to test it myself- I will be lazy now and thank you for the manual tests you've done :)

GabrielBrascher commented:
Thanks for testing, @slavkap. @andrijapanicsb you are right :-)

slavkap commented:
Thank you too @andrijapanicsb! I didn't know this one :)
The VR is recreated on the disabled zone/pod/cluster when I restart the network with those options:
allow.admin.vm.on.disabled.resources = false
allow.router.on.disabled.resources = true

slavkap commented:
yes, @DaanHoogland, it works as expected, and also I like this enhancement :)
